nvd/nvd.go

```diff
@@ -1,162 +1,250 @@
 		if err != nil {
-			return xerrors.Errorf("failed to decode NVD response: %w", err)
+			return xerrors.Errorf("unable to save entry CVEs for %q: %w", interval, err)
 		}
 
-		if err := save(nvd); err != nil {
-			return err
+		// if there are more records than maxResultsPerPage
+		// check next records page by page (first page is saved above)
+		for startIndex := u.maxResultsPerPage; startIndex < totalResults; startIndex += u.maxResultsPerPage {
+			_, err = u.saveEntry(interval, startIndex)
+			if err != nil {
+				return xerrors.Errorf("unable to save entry CVEs for: %s", interval)
+			}
 		}
-		bar.Increment()
 	}
-	bar.Finish()
+
 	return nil
 }
 
-func save(nvd *NVD) error {
-	for _, item := range nvd.CVEItems {
-		v, err := jsonpointer.Get(item, "/cve/CVE_data_meta/ID")
-		if err != nil {
-			log.Println(err)
-			continue
-		}
+func (u Updater) saveEntry(interval timeInterval, startIndex int) (int, error) {
+	entryURL, err := urlWithParams(u.baseURL, startIndex, u.maxResultsPerPage, interval)
+	if err != nil {
+		return 0, xerrors.Errorf("unable to get url with query parameters: %w", err)
+	}
 
-		cveID, ok := v.(string)
-		if !ok {
-			log.Println("failed to type assertion")
-			continue
-		}
+	var entry Entry
+	entry, err = u.getEntry(entryURL)
+	if err != nil {
+		return 0, xerrors.Errorf("unable to get entry for %q: %w", entryURL, err)
+	}
+	if err = save(entry, interval.lastModEndDate); err != nil {
+		return 0, xerrors.Errorf("unable to save entry: %w", err)
+	}
+	return entry.TotalResults, nil
+}
 
-		if err = utils.SaveCVEPerYear(filepath.Join(utils.VulnListDir(), feedDir), cveID, item); err != nil {
-			return xerrors.Errorf("failed to save NVD CVE detail: %w", err)
+func save(entry Entry, lastModEndDate string) error {
+	for _, cve := range entry.Vulnerabilities {
+		if err := utils.SaveCVEPerYear(filepath.Join(utils.VulnListDir(), apiDir), cve.Cve.ID, cve); err != nil {
+			return xerrors.Errorf("unable to write %s: %w", cve.Cve.ID, err)
 		}
 	}
+	// we need to save LastModEndDate to avoid saving wrong time if we get errors when saving all CVEs (first run)
+	// after this we can switch to entry.Timestamp
+	timestamp, err := time.Parse(nvdTimeFormat, lastModEndDate)
+	if err != nil {
+		return xerrors.Errorf("unable to parse timestamp: %w", err)
+	}
+	// update the Last_updated.json file after saving each entry
+	// to avoid overwriting this entry if we fail to save the next entry
+	err = utils.SetLastUpdatedDate("nvd", timestamp)
+	if err != nil {
+		return xerrors.Errorf("unable to save last_updated.json file: %w", err)
+	}
 	return nil
 }
 
-func fetchLastModifiedDate(feed string) (time.Time, error) {
-	log.Printf("Fetching NVD metadata(%s)...\n", feed)
-
-	url := fmt.Sprintf("%s/nvdcve-1.1-%s.meta", baseURL, feed)
-	res, err := utils.FetchURL(url, "", 5)
+func (u Updater) getEntry(url string) (Entry, error) {
+	var entry Entry
+	r, err := fetchURL(url, u.apiKey, u.retry)
 	if err != nil {
```

Comment:
We need to defer closing the body after checking the error

Reply:
Thanks! Changed in https://github.com/aquasecurity/vuln-list-update/pull/258/commits/763aac6a94019dd88489f0631e2ca63ee4ccd7fb